### DISCUSSION TITLE
RVE periodic boundary condition

### DISCUSSION THREAD

**Discussion by ada-ayy** [idaholab/moose#18367] [Q&A Modules: Solid mechanics] [answered]
Dear MOOSE，
I have a problem with the RVE periodic boundary condition: U^2-U^1=f(t), where the enforced displacement f(t) is applied on the opposite boundaries U^2 and U^1.  

I learn from #16419, then copy and modify the codes from `EqualValueBoundaryConstraint` and [https://github.com/crswong888/scorpion/blob/master/src/constraints/EqualValuePlusConstant.C](url)
This is my code:
[FunctionValueBoundaryConstraint-C.txt](https://github.com/idaholab/moose/files/6833992/FunctionValueBoundaryConstraint-C.txt)
[FunctionValueBoundaryConstraint-h.txt](https://github.com/idaholab/moose/files/6833996/FunctionValueBoundaryConstraint-h.txt)

However, I don't know exactly how to test the validity of the codes. But I find it always works like `DirichletBC`, as the nodes of U^1(U^2) always keep the same value. So are there any problems with my codes? Or other methods to achieve this boundary condition?

**Comment by hugary1995** [comment, marked as answer]
Well, if the values are equal, then it's clearly not working...

Also, did you see @SudiptaBiswas 's reply in your original post? https://github.com/idaholab/moose/discussions/16419#discussioncomment-198607

**Reply by ada-ayy** [reply]
Really thank you for your reply! But I am unfamiliar with the global strain system, and so I am still confused that how should I apply an increased strain on the whole model rather than the constant stress?
Thanks again!

**Reply by jbadger95** [reply]
I saw on your original post you were asking about using Lagrange multipliers to accomplish this. In case you're still interested, the answer is yes in 2D (although you'd need to modify the residual like suggested before) and almost in 3D. I have a PR up right now [#18203](https://github.com/idaholab/moose/pull/18203), I'm still putting the finishing touches on it but the code (in theory) is fully functional. 

If you modify the EqualValueConstraint (see `test/tests/mortar/solution-continuity/` for an example use of this constraint) and set `periodic = true`, it might work. For how to modify the residual, the continuity constraint is actually enforced when the LM variable acts as a test function. So if you think of the 2x2 block structure for the system, the constraint is enforced in the lower-left block and you'd essentially be adding a non-zero forcing term on the LM variable (so where you have (secondary - primary)*test you'd instead have (secondary - primary - your_value)*test ).

From a math perspective this gives me pause though, the solution will essentially be discontinuous and using a trace (the Lagrange multiplier) assumes you have enough regularity for it to be well-defined. Good luck!

**Reply by SudiptaBiswas** [reply]
@ada-ayy The global-strain system might just work for you. There are some examples here, https://github.com/idaholab/moose/tree/next/modules/combined/examples/periodic_strain on how to set this up. You can find several applications of how to utilize the system here, https://www.sciencedirect.com/science/article/pii/S0168874X20301165.
